src/cancellation_token.c

```diff
@@ -0,0 +1,212 @@
+        /*Codes_SRS_CANCELLATION_TOKEN_04_007: [ cancellation_token_is_canceled shall return a non-zero value if canceled is NULL. ]*/
+        LogError("Invalid args: THANDLE(CANCELLATION_TOKEN) cancellation_token = %p, canceled = %p", cancellation_token, canceled);
+        result = MU_FAILURE;
+    }
+    else
+    {
+        CANCELLATION_TOKEN* token_ptr = THANDLE_GET_T(CANCELLATION_TOKEN)(cancellation_token);
+        CANCELLATION_TOKEN_STATE state = interlocked_add(&token_ptr->state, 0);
+
+        /*Codes_SRS_CANCELLATION_TOKEN_04_008: [ cancellation_token_is_canceled shall assign true to *canceled if the token has been canceled and false otherwise. ]*/
+        if (state == CANCELLATION_TOKEN_STATE_CANCELED)
+        {
+            *canceled = true;
+        }
+        else
+        {
+            *canceled = false;
+        }
+
+        /*Codes_SRS_CANCELLATION_TOKEN_04_009: [ cancellation_token_is_canceled shall return 0 when successful. ]*/
+        result = 0;
+    }
+
+    return result;
+}
+
+IMPLEMENT_MOCKABLE_FUNCTION(, THANDLE(CANCELLATION_TOKEN_REGISTRATION), cancellation_token_register_notify, THANDLE(CANCELLATION_TOKEN), cancellation_token, TCALL_DISPATCHER_TARGET_FUNC_TYPE_NAME(CANCELLATION_TOKEN_CANCEL_CALL), on_cancel, void*, context)
+{
+    THANDLE(CANCELLATION_TOKEN_REGISTRATION) result = NULL;
+
+    if (cancellation_token == NULL || on_cancel == NULL)
+    {
+        /*Codes_SRS_CANCELLATION_TOKEN_04_010: [ cancellation_token_register_notify shall fail and return NULL when cancellation_token is NULL. ]*/
+        /*Codes_SRS_CANCELLATION_TOKEN_04_011: [ cancellation_token_register_notify shall fail and return NULL when on_cancel is NULL. ]*/
+        LogError("Invalid args: THANDLE(CANCELLATION_TOKEN) cancellation_token = %p, TCALL_DISPATCHER_TARGET_FUNC_TYPE_NAME(CANCELLATION_TOKEN_CANCEL_CALL) on_cancel = %p, void* context = %p", cancellation_token, on_cancel, context);
+    }
+    else
+    {
+        CANCELLATION_TOKEN* token_ptr = THANDLE_GET_T(CANCELLATION_TOKEN)(cancellation_token);
+
+        /*Codes_SRS_CANCELLATION_TOKEN_04_013: [ cancellation_token_register_notify shall call TCALL_DISPATCHER_REGISTER_TARGET(CANCELLATION_TOKEN_CANCEL_CALL) to register on_cancel with the dispatcher. ]*/
+        TCALL_DISPATCHER_TARGET_HANDLE(CANCELLATION_TOKEN_CANCEL_CALL) cancel_notification_target_handle = TCALL_DISPATCHER_REGISTER_TARGET(CANCELLATION_TOKEN_CANCEL_CALL)(token_ptr->cancel_notification_call_dispatcher, on_cancel, context);
+        if (cancel_notification_target_handle == NULL)
+        {
+            /*Codes_SRS_CANCELLATION_TOKEN_04_012: [ cancellation_token_register_notify shall fail and return NULL when any underlying call fails. ]*/
+            LogError("TCALL_DISPATCHER_REGISTER_TARGET(CANCELLATION_TOKEN_CANCEL_CALL)(token_ptr->cancel_notification_call_dispatcher, on_cancel, context) failed.");
```

Comment:
>token_ptr->cancel_notification_call_dispatcher, on_cancel, context [](http://example.com/codeflow?start=87&length=66)

parameters can be listed with their values.

Reply:
Sure.